### DISCUSSION TITLE
Enabling Pagination on Protocol

### DISCUSSION THREAD

**Discussion by nitinmish** [beckn/protocol-specifications#328] [Core Specification]
BecKn works on stateless asynchronous APIs, whereby the search intent issued by a caller is responded with list of items matching the search intent and keywords. In context of ONDC, currently, a search intent is raised by a buyer platform, against a transaction ID. The Gateway assesses the search intent by domain and sub domain and broadcast the intent to relevant platforms to respond. 

Intent is further responded by respective seller platforms pushing catalogs on the same transaction ID to ensure session sticky ness for the search responses. We have observed that in some cases, the number of items in a search response array is significantly large leading to processing bottlenecks. Thus, a pagination feature is needed to enable seller platforms to respond against a single search intent. Following are inherent extension of the feature:

1. Pagination shall ensure the API transmission stays light weight and performant.
2. This will also ensure the buyer Apps to manage load on their side, leading to managing user experience better
3. Inherently, the catalog becomes more current and optimised for application of TTLs as the new search requests will mandatorily have to go for network search
4. Sellers will have more flexibility to order the catalog items basis their own requirements
5. Responses to buyer can be more efficient. 
6. In B2B context, same flag can be used to issue action request on BAP side to expect a final quote against a search query at a later time.

Following is a proposed flow with change in Context section of on_search payload:

* Introduction of a new flag in context with a booloen txn_open, on same transaction ID but distinct message ID

{
  "context": {
    "domain": "string",
    "country": "string",
    "region": "string",
    "action": "on_search",
    "version": "string",
    "bap_id": "string",
    "bap_uri": "string",
    "bpp_id": "string",
    "bpp_uri": "string",
    "transaction_id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
    "message_id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
    **“txn_open”: “Y”**
    "timestamp": "2022-11-28T08:49:27.973Z",
    "key": "string",
    "ttl": "string"
  }

Upon a search intent received by a buyer App, buyer app may keep the transaction open for a time specified in the API Specs contract. Buyer Apps can render the results on PDP as the search responses are received. Buyer App may extend the time-limit for closing the transaction if this flag is received in subsequent payload, capped at a maximum value. 

@core-wg-admin @pramodkvarma

**Comment by venkatramanm** [comment]
Pagination could make implementations more complex. Instead, what if we could send Max records in intent . No pagination. Instead consider refining the search query. 

Logic being we rarely go to page 2 or 3 even on a browser search engine results. 

Typically we remain in first page and keep refining the search.

Implementation having paging can get quite complex . In. BPP, based on sort , a complex start clause need be built each time based on last record returned. Since it's stateless, the last record needs to be passed from. Bap. 

Bap may also do sorting on its side from.presentaion standpoint. So BPP results must be retained as is in order to refire a search with a start criteria. 

This is more complex than it appears. 

Other participants views may help in arriving at a decision. 

---

 .

**Reply by pramodkvarma** [reply]
I tend to agree with @venkatramanm here. While pagination can happen on UX, pagination across systems is complex. Also, no users browse beyond 100-200 items, by which they tend to refine search. So, sending max record as a preferred param and/or BPP using a default max records based on domain context and saying "there are more records beyond max, please refine search" via an indicator flag (in BPP response) is most practical.

**Reply by nirmay** [reply]
@venkatramanm @pramodkvarma I agree with you completely on the complexity part.

if we look at the usecase for which the pagination is being proposed, it is that of caching by BAPs, essentially a catalog sync. This is a valid usecase and I believe there are other options we should consider to satisfy this usecase. Performing catalog sync over a search API will induce complexity as rightly pointed out here.

I would like to propose we consider having a separate API (say /sync) for an important reason that search is by design in the critical path of experience and a separate API would allow such sync to be performed in a cleaner manner (consider size, security etc.) than piggy backing on the semantics of /search API. This could be a separate discussion thread if we agree to go down this path.

**Reply by nitinmish** [reply]
I am drawing a parallel with the way TCP works. Post the establishment of a session after three way handshake, the data payload between the client and server (one requesting data and session and the other serving the same) transacts in chunks of data, for a window length decided at the time of handshake between transacting counter-parties. This ensures that during failures, while the session stays alive the transfer of data can continue from where it failed till the time RST is issued, and the connection moves to CLOSED_WAIT state. Session is closed once the server issues a FIN or an RST is received from either client or server. While the session management is responsibility on either side to give a synchronous experience, in reality, TCP works in an Async way, sending payload in sequences of byte lengths and receiving ACKs for each transaction. This ensures the experience on client  remains at the hand of higher level protocols such as WS, HTTP or TCP socket level implementations. 

In our case, it is also a session enabled through transaction IDs. Proposal is to enable chunks of data to be transmitted for a live session as long as either window length (in this case, timeout set by buyer apps to receive data payload) or the seller platform sends an explicit close transaction by way of txn_open, just like FIN is issued. 

Optimizing search is an absolute requirement, and I agree that search broadcast and responses to it must get optimized by way of algorithms set by either stakeholders limiting the large payloads getting transacted over wire, however, the digital maturity at either side of tech will never match, and therefore, protocol must provide for handles on either side to handle the response payloads in chunks. Window length may be left for buyer apps to decide as there are larger extensions of this same feature, such as request for quotations in B2B transactions, where by, a seller may respond back saying I may not have a good deal at this moment, however, I may send it after a set period of time by indicating to buyer platform that the trasnaction may still be kept open for a period till the time I work and get better quotations. 

Protocol must be enabled for chunking, and implementation may be left with either parties on how they want to consume the same. The buyer APP may decide to create an on-demand processing of the responses that may have been received in chunks in subsequent calls.

In most cases, in absence of txn_open flag, buyer Apps will close the session on an API contract. But wherever applicable, this extension may be used to satisfy a use case, keeping APIs performant on wire. 

Introducing a new API is also an overhead, as it will add one more end point maintenance with idiosyncratic implementations, while the same feature can be enabled through a minor tweak in existing framework. I agree, there will be one time complexity involved to implement and satisfy this particular use case, however, current systems may not need to make a change as the feature addition will also keep the APIs backward compatible. Ones not implementing will simply issue NACKs to following calls, once they have received first response against a transacting ID. 
@venkatramanm  @pramodkvarma @nirmay

**Reply by venkatramanm** [reply]
Catalog syncing currently seem to have two or three designs floating today. We can discuss that separately. 

1. A search provider provides one time XLS load, then active/inactive api that can be used to fine tune the catalog for search. 
    a. As a BPP plugin
    b. As a separate network participant

2. Gw to cache catalog, and respond to search from its cache. Also firing search on  bpps to update cached catalog based on onsearch. (Bhoumik's suggestion )

3. Bap side caching per day etc 

My personal preference is 1.b , 2

We can seed another thread if required with these options.

**Reply by venkatramanm** [reply]
@nitinmish , today protocol does support chunking till TTL ,via multiple on search responses albeit from different bpps ..

Today we may limiting one response to a BPP but is only a superficial one. We can  allow for a BPP (say sharded across geo regions ) to send one search for STD:022 and another for STD:080. (Or what ever else decided by BPP)

I don't have a problem to relax this restriction of one onsearch per search as long as it is controlled by BPP and not bap .. 

Bap has to accept all on search as additive and show in ui accordingly. 

This definitely makes sense.

@pramodkvarma @nirmay

**Reply by nirmay** [reply]
@nitinmish thanks, I now see how you are thinking about this.

I believe there is a lot of value in considering a combined approach.

Looking at how services on the internet perform (large and variable size) data transfer today, provider responds back to the consumer request by sharing a secured link of the file object with all the data rather than the data itself being sent over the wire. The response is async. The file is later downloaded directly by the consumer.

This approach has several obvious advantages, 
- less complexity on both sides managing the chunks, 
- use of standard protocols and libs for downloading files (we have hundreds of participants who will benefit from this)
- less data over the wire on HTTP as we are sharing only the link.

For Beckn, this approach has added advantage with the gateway no longer having to send back all responses. I think this is a significant win given additional hop on the network and already congested gateway having to manage all participant requests/callbacks.

So, why don't we consider a model where if catalog response on search is larger, then BPP responds back with a link to the catalog instead of chunking/splitting it?
I was initially thinking a new API will be needed to achieve this, but I see that the search semantics (like filter criteria) will be needed anyway and hence exploring this approach where benefits from both approaches are being combined.

@venkatramanm @pramodkvarma look forward to your feedback,
